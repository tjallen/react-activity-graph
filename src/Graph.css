.tooltip {
  pointer-events: none;
  color: #fff;
  background-color: #242424;
  width: 140px; 
  text-align: center;
  display: none;
  padding: 5px 4px;
  border-radius: 3px;
}
.root {
  width: 100%;
  background-color: #fff;
  margin: 0 auto;
  display: flex;
  justify-content: center;
}

.week {
  display: flex;
  flex-direction: column;
  height: auto;
  flex-shrink: 1;
  flex-grow: 0;
  flex-basis: 0;
}

.day {
  color: #999;
  height: 10px;
  width: 10px;
  text-align: center;
  margin: 1px;
}

.day:hover > .tooltip {
  display: block;
  position: relative;
  top: -32px;
  left: -70px;
  z-index: 99;
}