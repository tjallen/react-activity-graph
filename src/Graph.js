import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Week from './Week';
import {
  format,
  subDays,
  subWeeks,
  endOfWeek // eslint-disable-line comma-dangle
} from 'date-fns';
import isEqual from 'lodash.isequal';
import './Graph.css';

export default class Graph extends Component {
  constructor(props) {
    super(props);
    this.state = {
      weeks: this.createYear(this.props.data),
      maxValue: this.props.data.reduce((prev, curr) => {
        if (prev.value > curr.value) return prev.value;
        return curr.value;
      }, 0),
    };
    this.handleDayClick = this.handleDayClick.bind(this);
  }
  componentWillReceiveProps(nextProps) {
    console.log('cWRP', this.props, '=>', nextProps);
    if (!isEqual(nextProps.data, this.props.data)) {
      console.log('!equal, setState');
      this.setState({
        weeks: this.createYear(nextProps.data),
      });
    } else {
      console.log(this.props.data, '===', nextProps.data);
    }
  }
  // init all weeks into a year array
  createYear(data) {
    const { endDate, weekCount, leftToRight } = this.props;
    const rawDate = endOfWeek(endDate);
    const year = [];
    for (let i = 0; i < weekCount; i++) {
      const subbedDate = subWeeks(rawDate, i);
      const formattedDate = format(subbedDate, 'YYYY-MM-DD');
      if (leftToRight) {
        year.push(this.createWeek(data, formattedDate));
      } else {
        year.unshift(this.createWeek(data, formattedDate));
      }
    }
    return year;
  }
  handleDayClick(value, date, formattedDate) {
    const ret = {
      value,
      date,
      formattedDate,
    };
    this.props.onDayClick(ret);
  }
  // init weeks to push to the year array
  createWeek(data, startDate) {
    if (!data || !startDate) throw new Error('cWeek missing params');
    const week = [];
    for (let i = 0; i < 7; i++) {
      let value = this.props.nullValue;
      const rawDate = subDays(startDate, i);
      const date = format(rawDate, 'YYYY-MM-DD');
      const formattedDate = format(rawDate, 'ddd, MMM D, YYYY');
      // loop data from props & push to array if date matches
      data.forEach(t => {
        if (t.date === date) {
          value = t.value;
        }
      });
      const dayObj = {
        date,
        value,
        formattedDate,
      };
      if (this.props.leftToRight) {
        week.push(dayObj);
      } else {
        week.unshift(dayObj);
      }
    }
    return week;
  }
  render() {
    const styles = {
      display: 'inline-block',
      width: 'auto',
      margin: '5px',
    };
    return (
      <div style={styles} className="root">
        {this.state.weeks.map((week, index) =>
          <Week
            className="week"
            week={this.state.weeks[index]}
            key={index}
            color={this.props.color}
            bgColor={this.props.bgColor}
            maxValue={this.state.maxValue}
            onDayClick={this.handleDayClick}
          />
        )}
      </div>
    );
  }
}

Graph.propTypes = {
  data: PropTypes.arrayOf(PropTypes.object).isRequired,
  weekCount: PropTypes.number,
  color: PropTypes.string,
  bgColor: PropTypes.string,
  leftToRight: PropTypes.bool,
  endDate: PropTypes.oneOfType([
    PropTypes.instanceOf(Date),
    PropTypes.string, // TODO validator for YYYY-MM-DD
  ]),
  nullValue: PropTypes.any,
  onDayClick: PropTypes.func,
  debug: PropTypes.bool,
};

Graph.defaultProps = {
  weekCount: 53,
  color: '#008000',
  bgColor: '#EBEDF0',
  leftToRight: false,
  endDate: new Date(),
  nullValue: 0,
  onDayClick: function noOp() {},
};
