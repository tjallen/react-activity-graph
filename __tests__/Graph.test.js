import React from 'react';
import Graph from '../src/Graph';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';
configure({ adapter: new Adapter() });
import { format, subDays, subWeeks } from 'date-fns';

// const props = {
//   data: data,
//   weekCount: 53,
//   color: '#008000',
//   bgColor: '#EBEDF0',
//   leftToRight: false,
// };

const now = new Date();
const fDate = format(now, 'YYYY-MM-DD');
const fDateOneWeekAgo = format(subWeeks(fDate, 1), 'YYYY-MM-DD');
const fDateOneDayAgo = format(subDays(fDate, 1), 'YYYY-MM-DD');

const data = [
  { date: fDateOneWeekAgo, value: 1 },
  { date: fDateOneDayAgo, value: 2 },
  { date: fDate, value: 3 },
];

const component = renderer.create(
  <Graph data={data} weekCount={2} />
);

// const wrapper = shallow(
//   <Graph {...props} />
// );

test('Graph snapshot test', () => {
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
