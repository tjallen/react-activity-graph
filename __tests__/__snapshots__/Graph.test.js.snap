// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph snapshot test 1`] = `
<div
  style={
    Object {
      "display": "inline-block",
      "width": "auto",
    }
  }
>
  <div
    style={
      Object {
        "background": "#EBEDF0",
        "display": "inline-block",
        "float": "left",
      }
    }
  >
    <div
      className="sc-htpNat lcFUQH"
    >
      <div
        className="sc-bwzfXH dUywxE"
      >
        <div
          className="sc-bdVaJa dOxDA"
        >
          0
           on 
          Sun, Sep 17, 2017
        </div>
      </div>
    </div>
    <div
      className="sc-EHOje fqsyqh"
    >
      <div
        className="sc-ifAKCX hwRfnX"
      >
        <div
          className="sc-bxivhb gQiCQK"
        >
          0
           on 
          Mon, Sep 18, 2017
        </div>
      </div>
    </div>
    <div
      className="sc-htoDjs dpnqJM"
    >
      <div
        className="sc-gzVnrw ECQmS"
      >
        <div
          className="sc-bZQynM jkROJx"
        >
          0
           on 
          Tue, Sep 19, 2017
        </div>
      </div>
    </div>
    <div
      className="sc-gZMcBi gMNwBJ"
    >
      <div
        className="sc-iwsKbI dJFRKU"
      >
        <div
          className="sc-dnqmqq inxySn"
        >
          0
           on 
          Wed, Sep 20, 2017
        </div>
      </div>
    </div>
    <div
      className="sc-jTzLTM gXCKra"
    >
      <div
        className="sc-VigVT hbWLCn"
      >
        <div
          className="sc-gqjmRU fSvBYA"
        >
          1
           on 
          Thu, Sep 21, 2017
        </div>
      </div>
    </div>
    <div
      className="sc-cSHVUG fkgJpg"
    >
      <div
        className="sc-jzJRlG kCqztP"
      >
        <div
          className="sc-fjdhpX hsEdRP"
        >
          0
           on 
          Fri, Sep 22, 2017
        </div>
      </div>
    </div>
    <div
      className="sc-kgoBCf gYjrLW"
    >
      <div
        className="sc-chPdSV kYHtdZ"
      >
        <div
          className="sc-kAzzGY itfBWo"
        >
          0
           on 
          Sat, Sep 23, 2017
        </div>
      </div>
    </div>
  </div>
  <div
    style={
      Object {
        "background": "#EBEDF0",
        "display": "inline-block",
        "float": "left",
      }
    }
  >
    <div
      className="sc-dxgOiQ fqPnek"
    >
      <div
        className="sc-kpOJdX kPKelu"
      >
        <div
          className="sc-kGXeez cgkPFB"
        >
          0
           on 
          Sun, Sep 24, 2017
        </div>
      </div>
    </div>
    <div
      className="sc-eNQAEJ fzvpNt"
    >
      <div
        className="sc-jKJlTe dKmWsr"
      >
        <div
          className="sc-ckVGcZ hkQVfy"
        >
          0
           on 
          Mon, Sep 25, 2017
        </div>
      </div>
    </div>
    <div
      className="sc-kkGfuU gqspMv"
    >
      <div
        className="sc-kEYyzF jLlJoM"
      >
        <div
          className="sc-hMqMXs cOqzRQ"
        >
          0
           on 
          Tue, Sep 26, 2017
        </div>
      </div>
    </div>
    <div
      className="sc-eHgmQL eNzkDE"
    >
      <div
        className="sc-hSdWYo gIsFOb"
      >
        <div
          className="sc-iAyFgw jHykPN"
        >
          2
           on 
          Wed, Sep 27, 2017
        </div>
      </div>
    </div>
    <div
      className="sc-brqgnP dBWIOd"
    >
      <div
        className="sc-jWBwVP iDvLxM"
      >
        <div
          className="sc-cvbbAY MJlaz"
        >
          3
           on 
          Thu, Sep 28, 2017
        </div>
      </div>
    </div>
    <div
      className="sc-jDwBTQ iofttv"
    >
      <div
        className="sc-jAaTju joVjxG"
      >
        <div
          className="sc-cMljjf eqIYZw"
        >
          0
           on 
          Fri, Sep 29, 2017
        </div>
      </div>
    </div>
    <div
      className="sc-jlyJG hPwWfC"
    >
      <div
        className="sc-iRbamj wlxNh"
      >
        <div
          className="sc-gPEVay gbsfhR"
        >
          0
           on 
          Sat, Sep 30, 2017
        </div>
      </div>
    </div>
  </div>
</div>
`;
